src/ast/passes/semantic_analyser.cpp

```diff
@@ -2639,6 +2687,66 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
           << "trying to assign value of type '" << assignTy
           << "' when variable already contains a value of type '"
           << search->second << "'";
+    } else if (search->second.IsIntegerTy()) {
+      if (assignment.expr->is_literal) {
+        auto integer = static_cast<Integer *>(assignment.expr);
+        if (!search->second.IsEqual(assignTy)) {
+          int64_t value = integer->n;
+          bool can_fit = false;
+          if (integer->is_negative) {
+            if (!search->second.IsSigned()) {
+              LOG(ERROR, assignment.loc, err_)
+                  << "Type mismatch for " << var_ident << ": "
+                  << "trying to assign value of type '" << assignTy
+                  << "' when variable already contains a value of type '"
+                  << search->second << "'";
+              return;
+            } else {
+              auto min_max = getIntTypeRange(search->second);
+              can_fit = value >= min_max.first;
+            }
+          } else {
+            if (!search->second.IsSigned()) {
+              auto min_max = getUIntTypeRange(search->second);
+              can_fit = static_cast<uint64_t>(value) <= min_max.second;
+            } else {
+              auto min_max = getIntTypeRange(search->second);
+              can_fit = static_cast<uint64_t>(value) <=
+                        static_cast<uint64_t>(min_max.second);
```

Comment:
Why not like this?
```suggestion
              can_fit = value >= min_max.first && value <= min_max.second;
```

Reply:
We already know the value is non negative at this point and the lowest uint64 is 0.

Reply:
I must be missing something but I still don't understand why cast the values to `uint64_t` when this branch deals with signed ints.

Reply:
Sure, will add a comment.